From ca00212168a5b2f1006178794676acdc775d5d71 Mon Sep 17 00:00:00 2001
From: Jens Wiklander <jens.wiklander@linaro.org>
Date: Tue, 26 Apr 2016 11:18:26 +0200
Subject: [PATCH 1022/1695] libutee: add TUI interface and implementation

Adds Trusted UI interface and corresponding user space implementation.
A pseudo TA is defined for interaction with OP-TEE Core.

This change was picked from [1] and updated to match latest
OP-TEE OS coding style.

Upstream-Status: Inappropriate [ST downstream https://github.com/STMicroelectronics/optee_os.git 9c84f759cbe66694b26ec246f5198bf7e00bec73 libutee: add TUI interface and implementation]
Link: [1] https://github.com/jenswi-linaro/optee_os/tree/tui-old2
Co-developped-by: Jens Wiklander <jens.wiklander@linaro.org>
Signed-off-by: Jens Wiklander <jens.wiklander@linaro.org>
Signed-off-by: Lionel Debieve <lionel.debieve@foss.st.com>
Change-Id: Icde71264e520579da9503a77eb6c9f6d6a8488a1
Reviewed-on: https://gerrit.st.com/c/mpu/oe/optee/optee_os/+/391774
Reviewed-by: Thomas BOURGOIN <thomas.bourgoin@foss.st.com>
Reviewed-by: Yannick FERTRE <yannick.fertre@foss.st.com>
Tested-by: Yannick FERTRE <yannick.fertre@foss.st.com>
Domain-Review: Yannick FERTRE <yannick.fertre@foss.st.com>
---
 lib/libutee/include/pta_tui.h           |  56 +++
 lib/libutee/include/tee_tui_api.h       |  24 +
 lib/libutee/include/tee_tui_api_types.h | 119 +++++
 lib/libutee/tee_api_property.c          |  10 +-
 lib/libutee/tui/sub.mk                  |   5 +
 lib/libutee/tui/tee_tui.c               | 554 ++++++++++++++++++++++++
 lib/libutee/tui/tui_entry.c             | 502 +++++++++++++++++++++
 lib/libutee/tui/tui_keyboard.c          | 517 ++++++++++++++++++++++
 lib/libutee/tui/tui_private.h           | 107 +++++
 lib/libutee/tui/tui_pta.c               | 137 ++++++
 lib/libutee/tui/tui_widget.c            | 199 +++++++++
 11 files changed, 2228 insertions(+), 2 deletions(-)
 create mode 100644 lib/libutee/include/pta_tui.h
 create mode 100644 lib/libutee/include/tee_tui_api.h
 create mode 100644 lib/libutee/include/tee_tui_api_types.h
 create mode 100644 lib/libutee/tui/tee_tui.c
 create mode 100644 lib/libutee/tui/tui_entry.c
 create mode 100644 lib/libutee/tui/tui_keyboard.c
 create mode 100644 lib/libutee/tui/tui_private.h
 create mode 100644 lib/libutee/tui/tui_pta.c
 create mode 100644 lib/libutee/tui/tui_widget.c

diff --git a/lib/libutee/include/pta_tui.h b/lib/libutee/include/pta_tui.h
new file mode 100644
index 000000000..1efe2900e
--- /dev/null
+++ b/lib/libutee/include/pta_tui.h
@@ -0,0 +1,56 @@
+/* SPDX-License-Identifier: BSD-2-Clause */
+/*
+ * Copyright (c) 2016-2021, Linaro Limited
+ */
+
+#ifndef __PTA_TUI_H
+#define __PTA_TUI_H
+
+#define PTA_TUI_UUID	{ 0x9c199eb0, 0x4d2d, 0x41a5, { \
+			  0x8c, 0xe9, 0xf2, 0x08, 0x52, 0x15, 0x77, 0xd1 } }
+
+#define PTA_TUI_SCREEN_24BPP	0
+
+/*
+ * [out]  value[0].a:	width
+ * [out]  value[0].b:	height
+ * [out]  value[1].a:	width_dpi
+ * [out]  value[1].b:	height_dpi
+ * [out]  value[2].a:	bpp
+ * Returns TEE_Result
+ */
+#define PTA_TUI_GET_SCREEN_INFO		1
+
+/* See TEE_TUIInitSession() */
+#define PTA_TUI_INIT_SESSION		3
+
+/* See TEE_TUICloseSession() */
+#define PTA_TUI_CLOSE_SESSION		4
+
+/*
+ * [in] value[0].a:	color
+ */
+#define PTA_TUI_INIT_SCREEN		5
+
+/*
+ * [in] value[0].a:	xpos
+ * [in] value[0].b:	ypos
+ * [in] value[1].a:	width
+ * [in] value[1].b:	height
+ * [in] memref[2]:	image
+ */
+#define PTA_TUI_SET_SCREEN_IMAGE	6
+
+/*
+ * See TEE_TUIDisplayScreen()
+ * [in]  value[0].a	clear_input	(bool)
+ * [in]  value[0].b	time_out	(milliseconds)
+ * [out] value[1].a	is_timedout	(bool)
+ * [out] value[2].a	is_key		(bool)
+ * [out] value[2].b	key
+ * [out] value[3].a	xpos
+ * [out] value[3].b	ypos
+ */
+#define PTA_TUI_DISPLAY_SCREEN		7
+
+#endif /*__PTA_TUI_H*/
diff --git a/lib/libutee/include/tee_tui_api.h b/lib/libutee/include/tee_tui_api.h
new file mode 100644
index 000000000..659433a50
--- /dev/null
+++ b/lib/libutee/include/tee_tui_api.h
@@ -0,0 +1,24 @@
+/* SPDX-License-Identifier: BSD-2-Clause */
+/*
+ * Copyright (c) 2016-2021, Linaro Limited
+ */
+
+#ifndef __TEE_TUI_API_H
+#define __TEE_TUI_API_H
+
+#include <tee_tui_api_types.h>
+
+TEE_Result TEE_TUICheckTextFormat(const char *text, uint32_t *width,
+				  uint32_t *height, uint32_t *lastIndex);
+TEE_Result TEE_TUIGetScreenInfo(TEE_TUIScreenOrientation screenOrientation,
+				uint32_t nbEntryFields,
+				TEE_TUIScreenInfo *screenInfo);
+TEE_Result TEE_TUIInitSession(void);
+TEE_Result TEE_TUICloseSession(void);
+TEE_Result TEE_TUIDisplayScreen(const TEE_TUIScreenConfiguration *screenConfiguration,
+				bool closeTUISession,
+				TEE_TUIEntryField *entryFields,
+				uint32_t entryFieldCount,
+				TEE_TUIButtonType *selectedButton);
+
+#endif /*__TEE_TUI_API_H*/
diff --git a/lib/libutee/include/tee_tui_api_types.h b/lib/libutee/include/tee_tui_api_types.h
new file mode 100644
index 000000000..78f56e110
--- /dev/null
+++ b/lib/libutee/include/tee_tui_api_types.h
@@ -0,0 +1,119 @@
+/* SPDX-License-Identifier: BSD-2-Clause */
+/*
+ * Copyright (c) 2016-2021, Linaro Limited
+ */
+
+#ifndef __TEE_TUI_API_TYPES_H
+#define __TEE_TUI_API_TYPES_H
+
+#include <types_ext.h>
+
+#define TEE_TUI_NUMBER_BUTTON_TYPES	0x00000006
+
+typedef enum {
+	TEE_TUI_HIDDEN_MODE = 0,
+	TEE_TUI_CLEAR_MODE,
+	TEE_TUI_TEMPORARY_CLEAR_MODE,
+} TEE_TUIEntryFieldMode;
+
+typedef enum {
+	TEE_TUI_NUMERICAL = 0,
+	TEE_TUI_ALPHANUMERICAL,
+} TEE_TUIEntryFieldType;
+
+typedef enum {
+	TEE_TUI_PORTRAIT = 0,
+	TEE_TUI_LANDSCAPE,
+} TEE_TUIScreenOrientation;
+
+typedef enum {
+	TEE_TUI_CORRECTION = 0,
+	TEE_TUI_OK,
+	TEE_TUI_CANCEL,
+	TEE_TUI_VALIDATE,
+	TEE_TUI_PREVIOUS,
+	TEE_TUI_NEXT,
+} TEE_TUIButtonType;
+
+typedef enum {
+	TEE_TUI_NO_SOURCE = 0,
+	TEE_TUI_REF_SOURCE,
+	TEE_TUI_OBJECT_SOURCE,
+} TEE_TUIImageSource;
+
+typedef struct {
+	TEE_TUIImageSource source;
+
+	__extension__ union {
+		struct {
+			void *image;
+			size_t imageLength;
+		} ref;
+		struct {
+			uint32_t storageID;
+			void *objectID;
+			size_t objectIDLen;
+		} object;
+	};
+	uint32_t width;
+	uint32_t height;
+} TEE_TUIImage;
+
+typedef struct {
+	char *text;
+	uint32_t textXOffset;
+	uint32_t textYOffset;
+	uint8_t textColor[3];
+	TEE_TUIImage image;
+	uint32_t imageXOffset;
+	uint32_t imageYOffset;
+} TEE_TUIScreenLabel;
+
+typedef struct {
+	char *text;
+	TEE_TUIImage image;
+} TEE_TUIButton;
+
+typedef struct {
+	TEE_TUIScreenOrientation screenOrientation;
+	TEE_TUIScreenLabel label;
+	TEE_TUIButton * buttons[TEE_TUI_NUMBER_BUTTON_TYPES];
+	bool requestedButtons[TEE_TUI_NUMBER_BUTTON_TYPES];
+} TEE_TUIScreenConfiguration;
+
+typedef struct {
+	char *buttonText;
+	uint32_t buttonWidth;
+	uint32_t buttonHeight;
+	bool buttonTextCustom;
+	bool buttonImageCustom;
+} TEE_TUIScreenButtonInfo;
+
+typedef struct {
+	uint32_t grayscaleBitsDepth;
+	uint32_t redBitsDepth;
+	uint32_t greenBitsDepth;
+	uint32_t blueBitsDepth;
+	uint32_t widthInch;
+	uint32_t heightInch;
+	uint32_t maxEntryFields;
+	uint32_t entryFieldLabelWidth;
+	uint32_t entryFieldLabelHeight;
+	uint32_t maxEntryFieldLength;
+	uint8_t labelColor[3];
+	uint32_t labelWidth;
+	uint32_t labelHeight;
+	TEE_TUIScreenButtonInfo buttonInfo[TEE_TUI_NUMBER_BUTTON_TYPES];
+} TEE_TUIScreenInfo;
+
+typedef struct {
+	char *label;
+	TEE_TUIEntryFieldMode mode;
+	TEE_TUIEntryFieldType type;
+	uint32_t minExpectedLength;
+	uint32_t maxExpectedLength;
+	char *buffer;
+	size_t bufferLength;
+} TEE_TUIEntryField;
+
+#endif /*__TEE_TUI_API_TYPES_H*/
diff --git a/lib/libutee/tee_api_property.c b/lib/libutee/tee_api_property.c
index 5f33e2ca8..39606a7d6 100644
--- a/lib/libutee/tee_api_property.c
+++ b/lib/libutee/tee_api_property.c
@@ -10,6 +10,7 @@
 #include <tee_api_defines.h>
 #include <tee_api.h>
 #include <tee_api_types.h>
+#include <tee_tui_api_types.h>
 #include <tee_arith_internal.h>
 #include <tee_internal_api_extensions.h>
 #include <tee_isocket.h>
@@ -47,9 +48,14 @@ const struct user_ta_property tee_props[] = {
 		&(const uint32_t){TEE_ISOCKET_VERSION}
 	},
 	{
-		"gpd.tee.internalCore.version",
+		"gpd.tee.tui.languages",
+		USER_TA_PROP_TYPE_STRING,
+		"en"
+	},
+	{
+		"gpd.tee.tui.orientation",
 		USER_TA_PROP_TYPE_U32,
-		&(const uint32_t){TEE_CORE_API_VERSION}
+		&(const uint32_t){1 << TEE_TUI_LANDSCAPE}
 	},
 };
 
diff --git a/lib/libutee/tui/sub.mk b/lib/libutee/tui/sub.mk
index 58bb4dc62..68c555b43 100644
--- a/lib/libutee/tui/sub.mk
+++ b/lib/libutee/tui/sub.mk
@@ -1,3 +1,8 @@
+srcs-y += tee_tui.c
+srcs-y += tui_pta.c
+srcs-y += tui_widget.c
+srcs-y += tui_keyboard.c
+srcs-y += tui_entry.c
 srcs-y += utf8.c
 srcs-y += image.c
 srcs-y += image_png.c
diff --git a/lib/libutee/tui/tee_tui.c b/lib/libutee/tui/tee_tui.c
new file mode 100644
index 000000000..9af607d9e
--- /dev/null
+++ b/lib/libutee/tui/tee_tui.c
@@ -0,0 +1,554 @@
+// SPDX-License-Identifier: BSD-2-Clause
+/*
+ * Copyright (c) 2016-2021, Linaro Limited
+ */
+
+#include <sys/queue.h>
+#include <tee_api.h>
+#include <tee_tui_api.h>
+#include <util.h>
+#include <string.h>
+#include <stdlib.h>
+#include "font.h"
+#include "image.h"
+#include "tui_private.h"
+#include <assert.h>
+
+#define MAX_LABEL_ROWS		4
+
+/* update_button_info() has to be called before using this the first time */
+static TEE_TUIScreenButtonInfo button_info[TEE_TUI_NUMBER_BUTTON_TYPES] = {
+	[TEE_TUI_CORRECTION] = {
+			.buttonText = (char *)"Correction",
+			.buttonTextCustom = false,
+			.buttonImageCustom = true,
+		},
+	[TEE_TUI_OK] = {
+			.buttonText = (char *)"OK",
+			.buttonTextCustom = false,
+			.buttonImageCustom = true,
+		},
+	[TEE_TUI_CANCEL] = {
+			.buttonText = (char *)"Cancel",
+			.buttonTextCustom = false,
+			.buttonImageCustom = true,
+		},
+	[TEE_TUI_VALIDATE] = {
+			.buttonText = (char *)"Validate",
+			.buttonTextCustom = false,
+			.buttonImageCustom = true,
+		},
+	[TEE_TUI_PREVIOUS] = {
+			.buttonText = (char *)"Previous",
+			.buttonTextCustom = false,
+			.buttonImageCustom = true,
+		},
+	[TEE_TUI_NEXT] = {
+			.buttonText = (char *)"Next",
+			.buttonTextCustom = false,
+			.buttonImageCustom = true,
+		},
+};
+
+static bool check_result(TEE_Result res, const TEE_Result *results,
+			 size_t num_results)
+{
+	size_t n;
+
+	for (n = 0; n < num_results; n++)
+		if (res == results[n])
+			return true;
+	return false;
+}
+
+#define CHECK_RESULT(r, results) \
+	do { \
+		if (!check_result((r), (results), ARRAY_SIZE(results))) \
+			TEE_Panic(r); \
+	} while (0)
+
+TEE_Result TEE_TUICheckTextFormat(const char *text, uint32_t *width,
+				  uint32_t *height, uint32_t *lastIndex)
+{
+	TEE_Result res;
+	size_t w;
+	size_t h;
+	size_t last_index;
+
+	if (font_check_text_format(text, &w, &h, &last_index))
+		res = TEE_SUCCESS;
+	else
+		res = TEE_ERROR_NOT_SUPPORTED;
+	*width = w;
+	*height = h;
+	*lastIndex = last_index;
+	return res;
+}
+
+static TEE_Result update_button_info(void)
+{
+	size_t n;
+	size_t h;
+	size_t w;
+
+	for (n = 0; n < TEE_TUI_NUMBER_BUTTON_TYPES; n++) {
+		if (button_info[n].buttonWidth)
+			continue;
+
+		if (!font_check_text_format(button_info[n].buttonText, &w, &h,
+					    NULL))
+			return TEE_ERROR_GENERIC;
+		button_info[n].buttonWidth = w;
+		button_info[n].buttonHeight = h;
+	}
+
+	return TEE_SUCCESS;
+}
+
+TEE_Result TEE_TUIGetScreenInfo(TEE_TUIScreenOrientation screenOrientation,
+				uint32_t nbEntryFields,
+				TEE_TUIScreenInfo *screenInfo)
+{
+	TEE_Result res;
+	const TEE_Result results[] = { TEE_ERROR_NOT_SUPPORTED, TEE_SUCCESS, };
+	size_t width;
+	size_t height;
+	size_t width_dpi;
+	size_t height_dpi;
+	uint32_t bpp;
+	uint32_t label_color = COLOR_RGB_BLACK;
+
+	res = update_button_info();
+	if (res != TEE_SUCCESS)
+		TEE_Panic(res);
+	res = __tui_pta_get_screen_info(&width, &height, &width_dpi,
+					&height_dpi, &bpp);
+	CHECK_RESULT(res, results);
+	if (res != TEE_SUCCESS)
+		return res;
+	if (screenOrientation != TEE_TUI_LANDSCAPE)
+		res = TEE_ERROR_NOT_SUPPORTED;
+	if (nbEntryFields > MAX_ENTRY_FIELDS)
+		res = TEE_ERROR_NOT_SUPPORTED;
+	screenInfo->grayscaleBitsDepth = 255;
+	screenInfo->redBitsDepth = 255;
+	screenInfo->greenBitsDepth = 255;
+	screenInfo->blueBitsDepth = 255;
+	screenInfo->widthInch = width / width_dpi;
+	screenInfo->heightInch = height / height_dpi;
+	screenInfo->maxEntryFields = MAX_ENTRY_FIELDS;
+	screenInfo->entryFieldLabelWidth =
+		width - button_info[TEE_TUI_CORRECTION].buttonWidth;
+	screenInfo->entryFieldLabelHeight = font_get_text_height();
+	screenInfo->maxEntryFieldLength =
+		font_get_max_field_length(screenInfo->entryFieldLabelWidth -
+					  4 * BORDER_SIZE);
+	screenInfo->labelColor[0] = label_color;
+	screenInfo->labelColor[1] = label_color >> 8;
+	screenInfo->labelColor[2] = label_color >> 16;
+	screenInfo->labelWidth = width;
+	screenInfo->labelHeight = font_get_text_height() * MAX_LABEL_ROWS;
+	memcpy(screenInfo->buttonInfo, button_info, sizeof(button_info));
+	return res;
+}
+
+TEE_Result TEE_TUIInitSession(void)
+{
+	TEE_Result res;
+	const TEE_Result results[] = {
+		TEE_ERROR_BUSY, TEE_ERROR_OUT_OF_MEMORY, TEE_SUCCESS,
+	};
+
+	res = __tui_pta_init_session();
+	CHECK_RESULT(res, results);
+	return res;
+}
+
+TEE_Result TEE_TUICloseSession(void)
+{
+	TEE_Result res;
+	const TEE_Result results[] = {
+		TEE_ERROR_BUSY, TEE_ERROR_BAD_STATE, TEE_SUCCESS,
+	};
+
+	res = __tui_pta_close_session();
+	CHECK_RESULT(res, results);
+	return res;
+}
+
+static TEE_Result set_label_image(struct image *image, size_t x, size_t y,
+				  const TEE_TUIImage *tui_image)
+{
+	TEE_Result res = TEE_SUCCESS;
+	TEE_ObjectHandle handle = TEE_HANDLE_NULL;
+	void *buf = NULL;
+	TEE_ObjectInfo obj_info;
+	size_t count;
+
+	switch (tui_image->source) {
+	case TEE_TUI_REF_SOURCE:
+		if (!image_set_png(image, x, y, tui_image->ref.image,
+				   tui_image->ref.imageLength))
+			res = TEE_ERROR_BAD_FORMAT;
+		break;
+	case TEE_TUI_OBJECT_SOURCE:
+		res = TEE_OpenPersistentObject(tui_image->object.storageID,
+					       tui_image->object.objectID,
+					       tui_image->object.objectIDLen,
+					       TEE_DATA_FLAG_ACCESS_READ |
+					       TEE_DATA_FLAG_SHARE_READ,
+					       &handle);
+		if (res != TEE_SUCCESS)
+			break;
+
+		res = TEE_GetObjectInfo1(handle, &obj_info);
+		if (res != TEE_SUCCESS)
+			break;
+
+		buf = malloc(obj_info.dataSize);
+		if (!buf) {
+			res = TEE_ERROR_OUT_OF_MEMORY;
+			break;
+		}
+
+		res = TEE_ReadObjectData(handle, buf, obj_info.dataSize,
+					 &count);
+		if (res != TEE_SUCCESS)
+			break;
+
+		if (count != obj_info.dataSize) {
+			res = TEE_ERROR_BAD_FORMAT;
+			break;
+		}
+
+		if (!image_set_png(image, x, y, buf, obj_info.dataSize))
+			res = TEE_ERROR_BAD_FORMAT;
+		break;
+
+	default:
+		res = TEE_ERROR_BAD_FORMAT;
+	}
+
+	TEE_CloseObject(handle);
+	free(buf);
+	return res;
+}
+
+static TEE_Result set_label(const TEE_TUIScreenLabel *label, size_t *ypos)
+{
+	TEE_Result res;
+	struct image *image = NULL;
+	size_t w;
+	size_t h;
+	size_t xmin;
+	size_t ymin;
+	size_t xmax;
+	size_t ymax;
+	uint32_t color;
+
+	if (!font_check_text_format(label->text, &w, &h, NULL))
+		TEE_Panic(TEE_ERROR_NOT_SUPPORTED);
+
+	if (label->image.source != TEE_TUI_NO_SOURCE) {
+		xmin = MIN(label->imageXOffset, label->textXOffset);
+		ymin = MIN(label->imageYOffset, label->textYOffset);
+		xmax = MAX(label->imageXOffset + label->image.width,
+			   label->textXOffset + w);
+		ymax = MAX(label->imageYOffset + label->image.height,
+			   label->textYOffset + h);
+	} else {
+		xmin = label->textXOffset;
+		ymin = label->textYOffset;
+		xmax = label->textXOffset + w;
+		ymax = label->textYOffset + h;
+	}
+	ymin += *ypos;
+	ymax += *ypos;
+	*ypos = ymax;
+
+	image = image_alloc(xmax - xmin, ymax - ymin,
+			    COLOR_RGB_LIGHT_GOLDENROD);
+	if (!image)
+		return TEE_ERROR_OUT_OF_MEMORY;
+
+	if (label->image.source != TEE_TUI_NO_SOURCE) {
+		res = set_label_image(image, label->imageXOffset,
+				      label->imageYOffset, &label->image);
+		if (res != TEE_SUCCESS)
+			goto out;
+	}
+
+	color = (label->textColor[0] << 16) | (label->textColor[1] << 8) |
+		label->textColor[2];
+	if (!font_render_text(image, label->textYOffset - ymin,
+			      label->textXOffset - xmin, label->text, color)) {
+		res = TEE_ERROR_BAD_FORMAT;
+		goto out;
+	}
+
+	res = __tui_pta_set_screen_image(ymin, xmin, image);
+out:
+	image_free(image);
+	return res;
+}
+
+static TEE_Result create_buttons(struct tui_widget_head *twh, size_t scr_width,
+				 size_t scr_height,
+				 TEE_TUIButton * const *buttons,
+				 const bool req_buttons[TEE_TUI_NUMBER_BUTTON_TYPES])
+{
+	TEE_Result res;
+	size_t n;
+	uint32_t req = 0;
+	enum button_pos bpos[3];
+	TEE_TUIButtonType btypes[3];
+	size_t num_buttons;
+
+	for (n = 0; n < TEE_TUI_NUMBER_BUTTON_TYPES; n++)
+		req |= req_buttons[n] << n;
+
+	switch (req & ~(1 << TEE_TUI_CORRECTION)) {
+	case (1 << TEE_TUI_OK):
+		bpos[0] = POS_MIDDLE;
+		btypes[0] = TEE_TUI_OK;
+		num_buttons = 1;
+		break;
+	case (1 << TEE_TUI_OK) | (1 << TEE_TUI_PREVIOUS):
+		bpos[0] = POS_MIDDLE;
+		btypes[0] = TEE_TUI_OK;
+		bpos[1] = POS_LEFT;
+		btypes[1] = TEE_TUI_PREVIOUS;
+		num_buttons = 2;
+		break;
+	case (1 << TEE_TUI_CANCEL) | (1 << TEE_TUI_VALIDATE):
+		bpos[0] = POS_LEFT;
+		btypes[0] = TEE_TUI_CANCEL;
+		bpos[1] = POS_RIGHT;
+		btypes[1] = TEE_TUI_VALIDATE;
+		num_buttons = 2;
+		break;
+	case (1 << TEE_TUI_CANCEL) | (1 << TEE_TUI_NEXT):
+		bpos[0] = POS_LEFT;
+		btypes[0] = TEE_TUI_CANCEL;
+		bpos[1] = POS_RIGHT;
+		btypes[1] = TEE_TUI_NEXT;
+		num_buttons = 2;
+		break;
+	case (1 << TEE_TUI_CANCEL) | (1 << TEE_TUI_PREVIOUS) |
+	     (1 << TEE_TUI_NEXT):
+		bpos[0] = POS_MIDDLE;
+		btypes[0] = TEE_TUI_CANCEL;
+		bpos[1] = POS_LEFT;
+		btypes[1] = TEE_TUI_PREVIOUS;
+		bpos[2] = POS_RIGHT;
+		btypes[2] = TEE_TUI_NEXT;
+		num_buttons = 3;
+		break;
+	case (1 << TEE_TUI_CANCEL) | (1 << TEE_TUI_VALIDATE) |
+	     (1 << TEE_TUI_PREVIOUS):
+		bpos[0] = POS_MIDDLE;
+		btypes[0] = TEE_TUI_CANCEL;
+		bpos[1] = POS_LEFT;
+		btypes[1] = TEE_TUI_PREVIOUS;
+		bpos[2] = POS_RIGHT;
+		btypes[2] = TEE_TUI_VALIDATE;
+		num_buttons = 3;
+		break;
+	default:
+		return TEE_ERROR_NOT_SUPPORTED;
+	}
+
+	for (n = 0; n < num_buttons; n++) {
+		res = __tui_button_create(twh, btypes[n],
+					  button_info + btypes[n], buttons[n],
+					  bpos[n], scr_width, scr_height);
+		if (res != TEE_SUCCESS)
+			return res;
+	}
+	return TEE_SUCCESS;
+}
+
+static void twh_final(struct tui_widget_head *twh)
+{
+	while (!TAILQ_EMPTY(twh)) {
+		struct tui_widget *tw = TAILQ_FIRST(twh);
+
+		TAILQ_REMOVE(twh, tw, link);
+		tw->ops->destroy(tw);
+	}
+}
+
+static bool display_should_exit(struct tui_widget_head *twh,
+				struct tui_state *ts)
+{
+	if (!ts->clicked_button)
+		return false;
+
+	if (*ts->clicked_button == TEE_TUI_VALIDATE)
+		return __tui_entry_fields_are_ok(twh);
+
+	return true;
+}
+
+static TEE_Result display_screen(struct tui_widget_head *twh,
+				 TEE_TUIButtonType *selectedButton)
+{
+	TEE_Result res;
+	bool is_key;
+	bool is_timedout;
+	uint32_t key;
+	size_t timeout;
+	size_t xpos;
+	size_t ypos;
+	struct tui_widget *tw;
+	struct tui_state ts = { NULL };
+
+	/* Find keyboard and save pointer */
+	ts.keyboard = __tui_keyboard_find(twh);
+
+	/* Find first input and give focus */
+	tw = __tui_entry_field_find_first(twh);
+	if (tw) {
+		/* If there's no keyboard something is wrong */
+		if (!ts.keyboard)
+			return TEE_ERROR_GENERIC;
+		tw->ops->focus(tw, &ts);
+	}
+
+	/*
+	 * If there's no focus yet, select the first widget what can have
+	 * focus.
+	 */
+	TAILQ_FOREACH(tw, twh, link) {
+		if (!ts.focus)
+			tw->ops->focus(tw, &ts);
+	}
+
+	/* If no widget has focus by now something is wrong */
+	if (!ts.focus)
+		return TEE_ERROR_GENERIC;
+
+	while (!display_should_exit(twh, &ts)) {
+		TAILQ_FOREACH(tw, twh, link) {
+			res = tw->ops->paint(tw, &ts);
+			if (res != TEE_SUCCESS)
+				return res;
+		}
+		ts.clicked_button = NULL;
+		if (__tui_entry_field_has_temp_clear(ts.input))
+			timeout = 1000;
+		else
+			timeout = SIZE_MAX;
+		res = __tui_pta_display_screen(true, timeout, &is_timedout,
+					       &is_key, &key, &xpos, &ypos);
+		if (res != TEE_SUCCESS)
+			return res;
+
+		if (is_timedout) {
+			__tui_entry_field_reset_temp_clear(ts.input);
+			continue;
+		}
+
+		if (!is_key) {
+			tw = __tui_widget_find(twh, xpos, ypos);
+			if (!tw)
+				continue;
+
+			EMSG("(%zu,%zu) clicked widget at (%zu,%zu)(%zu,%zu)",
+			     xpos, ypos, tw->x, tw->y, tw->width, tw->height);
+			tw->ops->click(tw, &ts, xpos, ypos);
+			continue;
+		}
+
+		switch (key) {
+		case '\t':
+			tw = ts.focus;
+			tw->ops->unfocus(tw, &ts);
+			while (!ts.focus) {
+				tw = TAILQ_NEXT(tw, link);
+				if (!tw)
+					tw = TAILQ_FIRST(twh);
+				tw->ops->focus(tw, &ts);
+			}
+			break;
+		case '\n':
+		case '\r':
+			ts.focus->ops->click(ts.focus, &ts, -1, -1);
+			break;
+		default:
+			ts.focus->ops->input(ts.focus, &ts, key);
+			break;
+		}
+	}
+	if (!ts.clicked_button)
+		return TEE_ERROR_GENERIC;
+	*selectedButton = *ts.clicked_button;
+	return TEE_SUCCESS;
+}
+
+TEE_Result TEE_TUIDisplayScreen(const TEE_TUIScreenConfiguration *sc,
+				bool closeTUISession,
+				TEE_TUIEntryField *entryFields,
+				uint32_t entryFieldCount,
+				TEE_TUIButtonType *selectedButton)
+{
+	TEE_Result res;
+	size_t ypos = 0;
+	struct tui_widget_head twh = TAILQ_HEAD_INITIALIZER(twh);
+	size_t scr_height;
+	size_t scr_width;
+	const TEE_Result results[] = {
+		TEE_ERROR_OUT_OF_MEMORY, TEE_ERROR_ITEM_NOT_FOUND,
+		TEE_ERROR_ACCESS_CONFLICT, TEE_ERROR_BAD_FORMAT,
+		TEE_ERROR_BAD_STATE, TEE_ERROR_BUSY, TEE_ERROR_CANCEL,
+		TEE_ERROR_EXTERNAL_CANCEL, TEE_SUCCESS,
+	};
+
+	if (sc->screenOrientation != TEE_TUI_LANDSCAPE)
+		TEE_Panic(TEE_ERROR_NOT_SUPPORTED);
+
+	res = __tui_pta_get_screen_info(&scr_width, &scr_height,
+					NULL, NULL, NULL);
+	CHECK_RESULT(res, results);
+	if (res != TEE_SUCCESS)
+		return res;
+
+	res = __tui_pta_init_screen(COLOR_RGB_LIGHT_GRAY);
+	CHECK_RESULT(res, results);
+	if (res != TEE_SUCCESS)
+		goto out;
+
+	res = set_label(&sc->label, &ypos);
+	CHECK_RESULT(res, results);
+	if (res != TEE_SUCCESS)
+		goto out;
+
+	if (entryFieldCount) {
+		res = __tui_entry_fields_create(&twh, &ypos, scr_width,
+						button_info +
+						TEE_TUI_CORRECTION,
+						entryFields, entryFieldCount);
+		CHECK_RESULT(res, results);
+		if (res != TEE_SUCCESS)
+			goto out;
+
+		res = __tui_keyboard_create(&twh, &ypos, scr_width);
+		CHECK_RESULT(res, results);
+		if (res != TEE_SUCCESS)
+			goto out;
+	}
+
+	res = create_buttons(&twh, scr_width, scr_height, sc->buttons,
+			     sc->requestedButtons);
+	CHECK_RESULT(res, results);
+	if (res != TEE_SUCCESS)
+		goto out;
+
+	res = display_screen(&twh, selectedButton);
+
+out:
+	twh_final(&twh);
+	if (closeTUISession)
+		TEE_TUICloseSession();
+	return res;
+}
diff --git a/lib/libutee/tui/tui_entry.c b/lib/libutee/tui/tui_entry.c
new file mode 100644
index 000000000..09a3abb7f
--- /dev/null
+++ b/lib/libutee/tui/tui_entry.c
@@ -0,0 +1,502 @@
+// SPDX-License-Identifier: BSD-2-Clause
+/*
+ * Copyright (c) 2016-2021, Linaro Limited
+ */
+
+#include <tee_api.h>
+#include <tee_tui_api.h>
+#include <stdlib.h>
+#include <string.h>
+#include "font.h"
+#include "image.h"
+#include "tui_private.h"
+
+struct tui_correction {
+	struct tui_widget *input;
+	TEE_TUIScreenButtonInfo *bi;
+	struct tui_widget tw;
+};
+
+struct tui_input {
+	TEE_TUIEntryField *field;
+	struct tui_keyboard *keyboard;
+	size_t min_len;
+	size_t max_len;
+	bool temp_clear;
+	struct tui_widget tw;
+};
+
+struct field_geometry {
+	size_t input_width;
+	size_t input_height;
+	size_t input_x;
+	size_t input_yoffs;
+	size_t input_max_len;
+	size_t corr_width;
+	size_t corr_height;
+	size_t corr_x;
+	size_t corr_yoffs;
+	size_t max_height;
+};
+
+static bool __maybe_unused is_tui_correction(struct tui_widget *tw);
+static bool is_tui_input(struct tui_widget *tw);
+
+static struct tui_correction *to_tui_correction(struct tui_widget *tw)
+{
+	assert(is_tui_correction(tw));
+	return container_of(tw, struct tui_correction, tw);
+}
+
+static struct tui_input *to_tui_input(struct tui_widget *tw)
+{
+	assert(is_tui_input(tw));
+	return container_of(tw, struct tui_input, tw);
+}
+
+static void state_set_input(struct tui_state *ts, struct tui_widget *tw)
+{
+	if (ts->input != tw) {
+		ts->input = tw;
+		__tui_keyboard_select_layout(ts);
+	}
+}
+
+static TEE_Result op_correction_paint(struct tui_widget *tw,
+				      struct tui_state *ts)
+{
+	TEE_Result res;
+	struct image *image;
+	struct tui_correction *tc = to_tui_correction(tw);
+	char *text = tc->bi->buttonText;
+
+	if (!tw->updated)
+		return TEE_SUCCESS;
+
+	image = __tui_widget_image_alloc(tw, COLOR_RGB_LIGHT_GOLDENROD);
+	if (!image)
+		return TEE_ERROR_OUT_OF_MEMORY;
+
+	if (!font_render_text(image, 0, 0, text, COLOR_RGB_BLACK)) {
+		res = TEE_ERROR_BAD_FORMAT;
+		goto out;
+	}
+	res = __tui_widget_display(tw, ts, image);
+	if (res == TEE_SUCCESS)
+		tw->updated = false;
+out:
+	image_free(image);
+	return res;
+}
+
+static void op_correction_focus(struct tui_widget *tw,
+				struct tui_state *ts)
+{
+	struct tui_correction *tc = to_tui_correction(tw);
+
+	if (tw != ts->focus) {
+		tw->updated = true;
+		ts->focus = tw;
+		state_set_input(ts, tc->input);
+	}
+}
+
+static void op_correction_click(struct tui_widget *tw,
+				struct tui_state *ts, ssize_t x __unused,
+				ssize_t y __unused)
+{
+	if (tw != ts->focus) {
+		ts->focus->ops->unfocus(ts->focus, ts);
+		tw->ops->focus(tw, ts);
+	}
+
+	ts->input->ops->input(ts->input, ts, 0x08);
+}
+
+static void op_correction_destroy(struct tui_widget *tw)
+{
+	free(to_tui_correction(tw));
+}
+
+static const struct tui_widget_ops correction_ops = {
+	.paint = op_correction_paint,
+	.focus = op_correction_focus,
+	.unfocus = __tui_widget_op_unfocus,
+	.click = op_correction_click,
+	.input = __tui_widget_op_dummy_input,
+	.destroy = op_correction_destroy,
+};
+
+static bool is_tui_correction(struct tui_widget *tw)
+{
+	return tw->ops == &correction_ops;
+}
+
+static struct tui_correction *add_entry_correction(struct tui_widget_head *twh,
+						   struct field_geometry *fg,
+						   size_t ypos,
+						   TEE_TUIScreenButtonInfo *bi,
+						   struct tui_widget *input)
+{
+	struct tui_correction *twc = calloc(1, sizeof(*twc));
+
+	if (!twc)
+		return NULL;
+
+	twc->tw.ops = &correction_ops;
+	twc->tw.updated = true;
+	twc->bi = bi;
+	twc->input = input;
+	twc->tw.y = ypos + fg->corr_yoffs;
+	twc->tw.x = fg->corr_x;
+	twc->tw.width = fg->corr_width;
+	twc->tw.height = fg->corr_height;
+	__tui_widget_add(twh, &twc->tw);
+	return twc;
+}
+
+static TEE_Result op_input_paint(struct tui_widget *tw, struct tui_state *ts)
+{
+	TEE_Result res;
+	struct tui_input *ti = to_tui_input(tw);
+	char text[MAX_ENTRY_FIELD_LENGTH + 1];
+	size_t l = ti->field->bufferLength;
+	struct image *image;
+
+	if (!tw->updated)
+		return TEE_SUCCESS;
+
+	image = __tui_widget_image_alloc(tw, COLOR_RGB_LIGHT_GOLDENROD);
+	if (!image)
+		return TEE_ERROR_OUT_OF_MEMORY;
+
+	if (!l)
+		goto display;
+
+	switch (ti->field->mode) {
+	case TEE_TUI_TEMPORARY_CLEAR_MODE:
+		if (ti->temp_clear) {
+			memset(text, '*', l - 1);
+			text[l - 1] = ti->field->buffer[l - 1];
+			break;
+		}
+
+		/*FALLTHROUGH*/
+	case TEE_TUI_HIDDEN_MODE:
+		memset(text, '*', l);
+		break;
+	case TEE_TUI_CLEAR_MODE:
+		memcpy(text, ti->field->buffer, l);
+		break;
+	default:
+		res = TEE_ERROR_GENERIC;
+		goto out;
+	}
+	text[l] = '\0';
+
+	if (!font_render_text(image, 0, 0, text, COLOR_RGB_BLACK)) {
+		res = TEE_ERROR_BAD_FORMAT;
+		goto out;
+	}
+display:
+	res = __tui_widget_display(tw, ts, image);
+	if (res == TEE_SUCCESS)
+		tw->updated = false;
+out:
+	image_free(image);
+	return res;
+}
+
+static void op_input_focus(struct tui_widget *tw, struct tui_state *ts)
+{
+	if (tw != ts->focus) {
+		tw->updated = true;
+		ts->focus = tw;
+		state_set_input(ts, tw);
+	}
+}
+
+static void op_input_click(struct tui_widget *tw, struct tui_state *ts,
+			   ssize_t x __unused, ssize_t y __unused)
+{
+	if (tw != ts->focus) {
+		ts->focus->ops->unfocus(ts->focus, ts);
+		tw->ops->focus(tw, ts);
+	}
+}
+
+static bool op_input_input(struct tui_widget *tw,
+			   struct tui_state *ts __unused, uint32_t key)
+{
+	struct tui_input *ti = to_tui_input(tw);
+
+	if ((key == 0x7f || key == 0x08)) {
+		if (ti->field->bufferLength) {
+			ti->field->bufferLength--;
+			ti->field->buffer[ti->field->bufferLength] = '\0';
+			tw->updated = true;
+			return true;
+		}
+
+		EMSG("Input buffer empty, ignoring key %#" PRIX32, key);
+		return false;
+	}
+
+	if (ti->field->bufferLength >= ti->max_len) {
+		EMSG("Input buffer full, ignoring key %#" PRIX32, key);
+		return false;
+	}
+
+	if ((ti->field->type == TEE_TUI_ALPHANUMERICAL &&
+	     key >= 0x20 && key <= 0x7d) ||
+	    (ti->field->type == TEE_TUI_NUMERICAL &&
+	     key >= '0' && key <= '9')) {
+		ti->field->buffer[ti->field->bufferLength] = key;
+		ti->field->bufferLength++;
+		ti->temp_clear = true;
+		tw->updated = true;
+		return true;
+	}
+
+	EMSG("Ignoring invalid key %#" PRIX32, key);
+	return false;
+}
+
+static void op_input_destroy(struct tui_widget *tw)
+{
+	free(to_tui_input(tw));
+}
+
+TEE_TUIEntryFieldType __tui_entry_field_get_input_type(struct tui_widget *tw)
+{
+	return to_tui_input(tw)->field->type;
+}
+
+static const struct tui_widget_ops input_ops = {
+	.paint = op_input_paint,
+	.focus = op_input_focus,
+	.unfocus = __tui_widget_op_unfocus,
+	.click = op_input_click,
+	.input = op_input_input,
+	.destroy = op_input_destroy,
+};
+
+static bool is_tui_input(struct tui_widget *tw)
+{
+	return tw->ops == &input_ops;
+}
+
+static struct tui_input *add_entry_input(struct tui_widget_head *twh,
+					 struct field_geometry *fg, size_t ypos,
+					 TEE_TUIEntryField *ef)
+{
+	struct tui_input *twi = calloc(1, sizeof(*twi));
+
+	if (!twi)
+		return NULL;
+
+	twi->tw.ops = &input_ops;
+	twi->tw.updated = true;
+	twi->field = ef;
+	twi->min_len = ef->minExpectedLength;
+	if (ef->maxExpectedLength)
+		twi->max_len = ef->maxExpectedLength;
+	else
+		twi->max_len = fg->input_max_len;
+	twi->tw.y = ypos + fg->input_yoffs;
+	twi->tw.x = fg->input_x;
+	twi->tw.width = fg->input_width;
+	twi->tw.height = fg->input_height;
+	__tui_widget_add(twh, &twi->tw);
+	return twi;
+}
+
+static TEE_Result get_entry_field_size(TEE_TUIEntryField *efs, size_t num_efs,
+				       TEE_TUIScreenButtonInfo *cbi,
+				       size_t scr_width,
+				       struct field_geometry *fg)
+{
+	size_t max_label_width = 0;
+	size_t max_len = 0;
+	size_t max_width;
+	size_t n;
+	size_t l;
+
+	/*
+	 * An entry field looks like:
+	 * +------------------------------------------------------------+
+	 * | Label text ...                                             |
+	 * +------------------------------------+-----------------------+
+	 * | Input field                        | Correction button     |
+	 * +------------------------------------+-----------------------+
+	 * Each of the boxes in the pictures can have different heights and
+	 * width. The correction button should be on the same level (tw.y)
+	 * as the input field, this is accomplished by setting the center
+	 * line at the same y level.
+	 *
+	 * If there's multiple entry fields they should have the same width
+	 * to line up nicely, this is done by scanning the required maximum
+	 * width of each box in entry fields and pick that larges value for
+	 * each.
+	 */
+
+	fg->corr_width = cbi->buttonWidth;
+	fg->corr_height = cbi->buttonHeight;
+
+	fg->input_max_len = font_get_max_field_length(scr_width -
+						      fg->corr_width -
+						      4 * BORDER_SIZE);
+	for (n = 0; n < num_efs; n++) {
+		if (efs[n].maxExpectedLength > fg->input_max_len)
+			return TEE_ERROR_GENERIC;
+		l = efs[n].maxExpectedLength;
+		if (l > max_len)
+			max_len = l;
+
+		if (efs[n].label) {
+			if (!font_check_text_format(efs[n].label, &l,
+						    NULL, NULL))
+				return TEE_ERROR_GENERIC;
+			if (l > max_label_width)
+				max_label_width = l;
+		}
+	}
+
+	fg->input_width = font_get_max_field_width(max_len) + BORDER_SIZE * 2;
+	fg->input_height = font_get_text_height() + BORDER_SIZE * 2;
+
+	max_width = MAX(fg->input_width + fg->corr_width, max_label_width);
+	fg->max_height = MAX(fg->input_height, fg->corr_height);
+
+	if (max_width > scr_width)
+		return TEE_ERROR_GENERIC;
+
+	fg->input_x = (scr_width / 2) - (max_width / 2);
+	fg->corr_x = fg->input_x + fg->input_width;
+
+	fg->input_yoffs = (fg->max_height / 2) - (fg->input_height / 2);
+	fg->corr_yoffs = (fg->max_height / 2) - (fg->corr_height / 2);
+	return TEE_SUCCESS;
+}
+
+static TEE_Result set_entry_label(size_t xpos, size_t *ypos, const char *text)
+{
+	TEE_Result res;
+	struct image *image;
+	size_t w;
+	size_t h;
+
+	if (!text)
+		return TEE_SUCCESS;
+
+	if (!font_check_text_format(text, &w, &h, NULL))
+		return TEE_ERROR_GENERIC;
+
+	image = image_alloc(w, h, COLOR_RGB_YELLOW);
+	if (!image)
+		return TEE_ERROR_OUT_OF_MEMORY;
+
+	if (!font_render_text(image, 0, 0, text, COLOR_RGB_BLACK)) {
+		res = TEE_ERROR_GENERIC;
+		goto out;
+	}
+
+	res = __tui_pta_set_screen_image(xpos, *ypos, image);
+	if (res != TEE_SUCCESS)
+		goto out;
+
+	(*ypos) += h;
+out:
+	image_free(image);
+	return res;
+}
+
+TEE_Result __tui_entry_fields_create(struct tui_widget_head *twh,
+				     size_t *ypos, size_t scr_width,
+				     TEE_TUIScreenButtonInfo *cbi,
+				     TEE_TUIEntryField *efs, size_t num_efs)
+{
+	TEE_Result res;
+	struct field_geometry fg;
+	size_t yp = *ypos;
+	size_t n;
+
+	for (n = 0; n < num_efs; n++) {
+		if (!efs[n].maxExpectedLength)
+			efs[n].maxExpectedLength = efs[n].bufferLength;
+		efs[n].bufferLength = 0;
+	}
+
+	res = get_entry_field_size(efs, num_efs, cbi, scr_width,
+				   &fg);
+	if (res != TEE_SUCCESS)
+		return res;
+
+	for (n = 0; n < num_efs; n++) {
+		struct tui_input *twi;
+
+		res = set_entry_label(fg.input_x, &yp, efs[n].label);
+		if (res != TEE_SUCCESS)
+			return res;
+
+		twi = add_entry_input(twh, &fg, yp, efs + n);
+		if (!twi)
+			return TEE_ERROR_OUT_OF_MEMORY;
+
+		if (!add_entry_correction(twh, &fg, yp, cbi, &twi->tw))
+			return TEE_ERROR_OUT_OF_MEMORY;
+
+		yp += fg.max_height;
+	}
+
+	*ypos = yp;
+
+	return TEE_SUCCESS;
+}
+
+bool __tui_entry_fields_are_ok(struct tui_widget_head *twh)
+{
+	struct tui_widget *tw;
+
+	TAILQ_FOREACH(tw, twh, link) {
+		if (is_tui_input(tw)) {
+			struct tui_input *ti = to_tui_input(tw);
+
+			if (ti->field->bufferLength < ti->min_len)
+				return false;
+		}
+	}
+
+	return true;
+}
+
+bool __tui_entry_field_has_temp_clear(struct tui_widget *tw)
+{
+	if (tw) {
+		struct tui_input *ti = to_tui_input(tw);
+
+		return ti->temp_clear;
+	}
+
+	return false;
+}
+
+void __tui_entry_field_reset_temp_clear(struct tui_widget *tw)
+{
+	if (tw) {
+		struct tui_input *ti = to_tui_input(tw);
+
+		ti->temp_clear = false;
+		tw->updated = true;
+	}
+}
+
+struct tui_widget *__tui_entry_field_find_first(struct tui_widget_head *twh)
+{
+	struct tui_widget *tw;
+
+	TAILQ_FOREACH(tw, twh, link)
+		if (is_tui_input(tw))
+			return tw;
+	return NULL;
+}
diff --git a/lib/libutee/tui/tui_keyboard.c b/lib/libutee/tui/tui_keyboard.c
new file mode 100644
index 000000000..563d64f91
--- /dev/null
+++ b/lib/libutee/tui/tui_keyboard.c
@@ -0,0 +1,517 @@
+// SPDX-License-Identifier: BSD-2-Clause
+/*
+ * Copyright (c) 2016-2021, Linaro Limited
+ */
+
+#include <tee_api.h>
+#include <tee_tui_api.h>
+#include <stdlib.h>
+#include "font.h"
+#include "image.h"
+#include "tui_private.h"
+
+struct tui_key {
+	char value;
+	const char *text;
+	struct tui_widget tw;
+};
+
+struct tui_switchkey {
+	const char *text;
+	size_t layout_idx;
+	struct tui_widget tw;
+};
+
+struct tui_keyboard {
+	size_t layout_idx;
+	size_t num_layouts;
+	struct keyboard_head *kbh;
+	struct tui_widget tw;
+};
+
+struct keyboard_head {
+	TEE_TUIEntryFieldType type;
+	struct tui_widget_head twh;
+};
+
+struct keyboard_layout {
+	TEE_TUIEntryFieldType type;
+	const char *values;
+	size_t columns;
+	size_t rows;
+	const char *switch_str;
+	size_t switch_idx;
+};
+
+static const struct keyboard_layout keyboard_layouts[] = {
+	{
+		.type = TEE_TUI_ALPHANUMERICAL,
+		.values = (const char[]) {
+			'1', '2', '3', '4', '5', '6', '7', '8', '9', '0',
+			'q', 'w', 'e', 'r', 't', 'y', 'u', 'i', 'o', 'p',
+			'a', 's', 'd', 'f', 'g', 'h', 'j', 'k', 'l', 0,
+			'z', 'x', 'c', 'v', 'b', 'n', 'm', 0, 0, 0,
+		},
+		.columns = 10,
+		.rows = 4,
+		.switch_str = "<abc>",
+		.switch_idx = 1,
+	},
+	{
+		.type = TEE_TUI_ALPHANUMERICAL,
+		.values = (const char[]) {
+			'1', '2', '3', '4', '5', '6', '7', '8', '9', '0',
+			'Q', 'W', 'E', 'R', 'T', 'Y', 'U', 'I', 'O', 'P',
+			'A', 'S', 'D', 'F', 'G', 'H', 'J', 'K', 'L', 0,
+			'Z', 'X', 'C', 'V', 'B', 'N', 'M', 0, 0, 0,
+		},
+		.columns = 10,
+		.rows = 4,
+		.switch_str = "<ABC>",
+		.switch_idx = 2,
+	},
+	{
+		.type = TEE_TUI_ALPHANUMERICAL,
+		.values = (const char[]) {
+			'1', '2', '3', '4', '5', '6', '7', '8', '9', '0', 0,
+			'!', '@', '#', '$', '%', '^', '&', '*', '-', '+', '=',
+			',', '.', '?', ';', ':', '\'', '"', '`', '/', '\\', '_',
+			'(', ')', '<', '>', '[', ']', '{', '}', '|', 0, 0,
+		},
+		.columns = 11,
+		.rows = 4,
+		.switch_idx = 0,
+		.switch_str = "<!@#>",
+	},
+	{
+		.type = TEE_TUI_NUMERICAL,
+		.values = (const char []) {
+			'1', '2', '3',
+			'4', '5', '6',
+			'7', '8', '9',
+			 0, '0', 0,
+		},
+		.columns = 3,
+		.rows = 4,
+	},
+};
+
+static void op_dummy_unfocus(struct tui_widget *tw __unused,
+			     struct tui_state *ts __unused)
+{
+	/* Do nothing, this widget can't be in focus */
+}
+
+static void op_dummy_focus(struct tui_widget *tw __unused,
+			   struct tui_state *ts __unused)
+{
+	/* Do nothing, this widget can't be in focus */
+}
+
+static bool __maybe_unused is_tui_key(struct tui_widget *tw);
+static struct tui_key *to_tui_key(struct tui_widget *tw)
+{
+	assert(is_tui_key(tw));
+	return container_of(tw, struct tui_key, tw);
+}
+
+static bool __maybe_unused is_tui_switchkey(struct tui_widget *tw);
+static struct tui_switchkey *to_tui_switchkey(struct tui_widget *tw)
+{
+	assert(is_tui_switchkey(tw));
+	return container_of(tw, struct tui_switchkey, tw);
+}
+
+static bool is_tui_keyboard(struct tui_widget *tw);
+static struct tui_keyboard *to_tui_keyboard(struct tui_widget *tw)
+{
+	assert(is_tui_keyboard(tw));
+	return container_of(tw, struct tui_keyboard, tw);
+}
+
+static TEE_Result op_key_paint(struct tui_widget *tw, struct tui_state *ts)
+{
+	TEE_Result res;
+	struct tui_key *k = to_tui_key(tw);
+	struct image *image;
+	char text[2] = { k->value, '\0' };
+	const char *t;
+
+	if (k->text)
+		t = k->text;
+	else
+		t = text;
+
+	if (!tw->updated)
+		return TEE_SUCCESS;
+
+	image = __tui_widget_image_alloc(tw, COLOR_RGB_LIGHT_GOLDENROD);
+	if (!image)
+		return TEE_ERROR_OUT_OF_MEMORY;
+
+	if (!font_render_text(image, 0, 0, t, COLOR_RGB_BLACK)) {
+		res = TEE_ERROR_BAD_FORMAT;
+		goto out;
+	}
+	res = __tui_widget_display(tw, ts, image);
+	if (res == TEE_SUCCESS)
+		tw->updated = false;
+out:
+	image_free(image);
+	return res;
+}
+
+static void op_key_click(struct tui_widget *tw, struct tui_state *ts,
+			 ssize_t x __unused, ssize_t y __unused)
+{
+	struct tui_key *k = to_tui_key(tw);
+
+	ts->input->ops->input(ts->input, ts, k->value);
+}
+
+static void op_key_destroy(struct tui_widget *tw)
+{
+	free(to_tui_key(tw));
+}
+
+static const struct tui_widget_ops key_ops = {
+	.paint = op_key_paint,
+	.focus = op_dummy_focus,
+	.unfocus = op_dummy_unfocus,
+	.click = op_key_click,
+	.input = __tui_widget_op_dummy_input,
+	.destroy = op_key_destroy,
+};
+
+static bool is_tui_key(struct tui_widget *tw)
+{
+	return tw->ops == &key_ops;
+}
+
+static TEE_Result add_key(struct tui_widget_head *twh, char value,
+			  const char *text, size_t x, size_t y,
+			  size_t w, size_t h)
+{
+	struct tui_key *tk = calloc(1, sizeof(*tk));
+
+	if (!tk)
+		return TEE_ERROR_OUT_OF_MEMORY;
+	tk->tw.ops = &key_ops;
+	tk->tw.updated = true;
+	tk->tw.x = x;
+	tk->tw.y = y;
+	tk->tw.width = w;
+	tk->tw.height = h;
+	tk->value = value;
+	tk->text = text;
+	TAILQ_INSERT_TAIL(twh, &tk->tw, link);
+	return TEE_SUCCESS;
+}
+
+static TEE_Result op_switchkey_paint(struct tui_widget *tw,
+				     struct tui_state *ts)
+{
+	TEE_Result res;
+	struct tui_switchkey *k = to_tui_switchkey(tw);
+	struct image *image;
+
+	if (!tw->updated)
+		return TEE_SUCCESS;
+
+	image = __tui_widget_image_alloc(tw, COLOR_RGB_LIGHT_GOLDENROD);
+	if (!image)
+		return TEE_ERROR_OUT_OF_MEMORY;
+
+	if (!font_render_text(image, 0, 0, k->text, COLOR_RGB_BLACK)) {
+		res = TEE_ERROR_BAD_FORMAT;
+		goto out;
+	}
+	res = __tui_widget_display(tw, ts, image);
+	if (res == TEE_SUCCESS)
+		tw->updated = false;
+out:
+	image_free(image);
+	return res;
+}
+
+static void op_switchkey_click(struct tui_widget *tw, struct tui_state *ts,
+			       ssize_t x __unused, ssize_t y __unused)
+{
+	struct tui_switchkey *k = to_tui_switchkey(tw);
+	struct tui_keyboard *kb = to_tui_keyboard(ts->keyboard);
+
+	kb->layout_idx = k->layout_idx;
+	ts->keyboard->updated = true;
+}
+
+static void op_switchkey_destroy(struct tui_widget *tw)
+{
+	free(to_tui_switchkey(tw));
+}
+
+static const struct tui_widget_ops switchkey_ops = {
+	.paint = op_switchkey_paint,
+	.focus = op_dummy_focus,
+	.unfocus = op_dummy_unfocus,
+	.click = op_switchkey_click,
+	.input = __tui_widget_op_dummy_input,
+	.destroy = op_switchkey_destroy,
+};
+
+static bool is_tui_switchkey(struct tui_widget *tw)
+{
+	return tw->ops == &switchkey_ops;
+}
+
+static TEE_Result add_switchkey(struct tui_widget_head *twh,
+				const char *text, size_t layout_idx,
+				size_t x, size_t y)
+{
+	struct tui_switchkey *k;
+	size_t w;
+	size_t h;
+
+	if (!font_check_text_format(text, &w, &h, NULL))
+		return TEE_ERROR_GENERIC;
+
+	k = calloc(1, sizeof(*k));
+	if (!k)
+		return TEE_ERROR_OUT_OF_MEMORY;
+	k->tw.ops = &switchkey_ops;
+	k->tw.updated = true;
+	k->tw.x = x;
+	k->tw.y = y;
+	k->tw.width = w;
+	k->tw.height = h;
+	k->text = text;
+	k->layout_idx = layout_idx;
+	TAILQ_INSERT_TAIL(twh, &k->tw, link);
+	return TEE_SUCCESS;
+}
+
+static TEE_Result op_keyboard_paint(struct tui_widget *tw, struct tui_state *ts)
+{
+	TEE_Result res;
+	struct tui_keyboard *tkb = to_tui_keyboard(tw);
+	struct tui_widget *tw_key;
+	struct image *image;
+
+	if (!tw->updated)
+		return TEE_SUCCESS;
+
+	image = __tui_widget_image_alloc(tw, COLOR_RGB_LIGHT_GRAY);
+	if (!image)
+		return TEE_ERROR_OUT_OF_MEMORY;
+	res = __tui_widget_display(tw, ts, image);
+	image_free(image);
+	if (res != TEE_SUCCESS)
+		return res;
+
+	TAILQ_FOREACH(tw_key, &tkb->kbh[tkb->layout_idx].twh, link) {
+		tw_key->updated = true;
+		res = tw_key->ops->paint(tw_key, ts);
+		if (res != TEE_SUCCESS)
+			return res;
+	}
+	tw->updated = false;
+	return TEE_SUCCESS;
+}
+
+static void op_keyboard_click(struct tui_widget *tw, struct tui_state *ts,
+			      ssize_t x, ssize_t y)
+{
+	struct tui_keyboard *tkb = to_tui_keyboard(tw);
+	struct tui_widget *tw_key;
+
+	tw_key = __tui_widget_find(&tkb->kbh[tkb->layout_idx].twh, x, y);
+	if (tw_key)
+		tw_key->ops->click(tw_key, ts, x, y);
+}
+
+static void op_keyboard_destroy(struct tui_widget *tw)
+{
+	struct tui_keyboard *tkb = to_tui_keyboard(tw);
+	struct tui_widget *tw_key;
+	size_t n;
+
+	for (n = 0; n < tkb->num_layouts; n++) {
+		while (true) {
+			tw_key = TAILQ_FIRST(&tkb->kbh[n].twh);
+			if (!tw_key)
+				break;
+			TAILQ_REMOVE(&tkb->kbh[n].twh, tw_key, link);
+			tw_key->ops->destroy(tw_key);
+		}
+	}
+	free(tkb->kbh);
+	free(tkb);
+}
+
+static const struct tui_widget_ops keyboard_ops = {
+	.paint = op_keyboard_paint,
+	.focus = op_dummy_focus,
+	.unfocus = op_dummy_unfocus,
+	.click = op_keyboard_click,
+	.input = __tui_widget_op_dummy_input,
+	.destroy = op_keyboard_destroy,
+};
+
+static bool is_tui_keyboard(struct tui_widget *tw)
+{
+	return tw->ops == &keyboard_ops;
+}
+
+TEE_Result __tui_keyboard_create(struct tui_widget_head *twh, size_t *ypos,
+				 size_t scr_width)
+{
+	TEE_Result res;
+	struct tui_keyboard *tkb;
+	size_t row;
+	size_t col;
+	size_t hk_max = 0;
+	size_t wk_max = 0;
+	size_t hmax = 0;
+	size_t wmax = 0;
+	size_t h;
+	size_t w;
+	size_t n;
+
+	for (n = 0; n < ARRAY_SIZE(keyboard_layouts); n++) {
+		for (row = 0; row < keyboard_layouts[n].rows; row++) {
+			for (col = 0; col < keyboard_layouts[n].columns;
+			     col++) {
+				char text[2] = {
+					keyboard_layouts[n].values[row *
+						keyboard_layouts[n].columns +
+						col],
+					'\0',
+				 };
+
+				if (!text[0])
+					continue;
+				if (!font_check_text_format(text, &w, &h, NULL))
+					return TEE_ERROR_GENERIC;
+				if (w > wk_max)
+					wk_max = w;
+				if (h > hk_max)
+					hk_max = h;
+			}
+		}
+	}
+
+	for (n = 0; n < ARRAY_SIZE(keyboard_layouts); n++) {
+		size_t num_rows = keyboard_layouts[n].rows;
+
+		if (keyboard_layouts[n].type == TEE_TUI_ALPHANUMERICAL)
+			num_rows++;
+		hmax = MAX(num_rows * hk_max, hmax);
+		wmax = MAX(keyboard_layouts[n].columns * wk_max, wmax);
+	}
+
+	tkb = calloc(1, sizeof(*tkb));
+	if (!tkb)
+		return TEE_ERROR_OUT_OF_MEMORY;
+
+	tkb->kbh = calloc(ARRAY_SIZE(keyboard_layouts), sizeof(*tkb->kbh));
+	if (!tkb->kbh) {
+		res = TEE_ERROR_OUT_OF_MEMORY;
+		goto out;
+	}
+
+	for (n = 0; n < ARRAY_SIZE(keyboard_layouts); n++)
+		TAILQ_INIT(&tkb->kbh[n].twh);
+
+	tkb->tw.ops = &keyboard_ops;
+	tkb->tw.updated = true;
+	tkb->tw.width = wmax;
+	tkb->tw.height = hmax;
+	tkb->num_layouts = ARRAY_SIZE(keyboard_layouts);
+
+	if (tkb->tw.width > scr_width) {
+		res = TEE_ERROR_GENERIC;
+		goto out;
+	}
+	tkb->tw.x = (scr_width - tkb->tw.width) / 2;
+	tkb->tw.y = *ypos;
+
+	for (n = 0; n < ARRAY_SIZE(keyboard_layouts); n++) {
+		size_t xoffs;
+
+		xoffs = (wmax - (keyboard_layouts[n].columns * wk_max)) / 2;
+		tkb->kbh[n].type = keyboard_layouts[n].type;
+
+		for (row = 0; row < keyboard_layouts[n].rows; row++) {
+			for (col = 0; col < keyboard_layouts[n].columns;
+			     col++) {
+				char value = keyboard_layouts[n].values[row *
+					keyboard_layouts[n].columns +
+					col];
+				if (!value)
+					continue;
+				res = add_key(&tkb->kbh[n].twh, value, NULL,
+					      tkb->tw.x + col * wk_max + xoffs,
+					      tkb->tw.y + row * hk_max,
+					      wk_max, hk_max);
+				if (res != TEE_SUCCESS)
+					goto out;
+			}
+		}
+
+		if (keyboard_layouts[n].type == TEE_TUI_ALPHANUMERICAL) {
+			const char *str = "<space>";
+			size_t switch_idx = keyboard_layouts[n].switch_idx;
+
+			res = add_switchkey(&tkb->kbh[n].twh,
+					    keyboard_layouts[switch_idx].switch_str,
+					    switch_idx, tkb->tw.x + xoffs,
+					    tkb->tw.y +
+					    keyboard_layouts[n].rows * hk_max);
+			if (res != TEE_SUCCESS)
+				goto out;
+
+			if (!font_check_text_format(str, &w, NULL, NULL)) {
+				res = TEE_ERROR_GENERIC;
+				goto out;
+			}
+
+			res = add_key(&tkb->kbh[n].twh, ' ', str,
+				      tkb->tw.x + 5 * wk_max + xoffs,
+				      tkb->tw.y +
+					keyboard_layouts[n].rows * hk_max,
+				      w, hk_max);
+			if (res != TEE_SUCCESS)
+				goto out;
+		}
+	}
+
+	TAILQ_INSERT_TAIL(twh, &tkb->tw, link);
+out:
+	if (res != TEE_SUCCESS)
+		op_keyboard_destroy(&tkb->tw);
+	return res;
+}
+
+void __tui_keyboard_select_layout(struct tui_state *ts)
+{
+	struct tui_keyboard *kb = to_tui_keyboard(ts->keyboard);
+	const TEE_TUIEntryFieldType type =
+		__tui_entry_field_get_input_type(ts->input);
+
+	if (type == kb->kbh[kb->layout_idx].type)
+		return;
+	kb->layout_idx = 0;
+	kb->tw.updated = true;
+
+	while (type != kb->kbh[kb->layout_idx].type)
+		kb->layout_idx = (kb->layout_idx + 1) % kb->num_layouts;
+}
+
+struct tui_widget *__tui_keyboard_find(struct tui_widget_head *twh)
+{
+	struct tui_widget *tw;
+
+	TAILQ_FOREACH(tw, twh, link)
+		if (is_tui_keyboard(tw))
+			return tw;
+	return NULL;
+}
diff --git a/lib/libutee/tui/tui_private.h b/lib/libutee/tui/tui_private.h
new file mode 100644
index 000000000..2b5a0d3b5
--- /dev/null
+++ b/lib/libutee/tui/tui_private.h
@@ -0,0 +1,107 @@
+/* SPDX-License-Identifier: BSD-2-Clause */
+/*
+ * Copyright (c) 2016-2021, Linaro Limited
+ */
+
+#ifndef __TUI_PRIVATE_H
+#define __TUI_PRIVATE_H
+
+#include <sys/queue.h>
+#include <tee_api.h>
+#include <tee_tui_api.h>
+#include <types_ext.h>
+#include <util.h>
+#include <assert.h>
+#include "image.h"
+
+#define MAX_ENTRY_FIELDS	2
+#define MAX_ENTRY_FIELD_LENGTH	32
+#define BORDER_SIZE		2
+
+enum button_pos {
+	POS_LEFT,
+	POS_MIDDLE,
+	POS_RIGHT,
+};
+
+struct tui_state {
+	TEE_TUIButtonType *clicked_button;
+	struct tui_widget *focus;
+	struct tui_widget *input;
+	struct tui_widget *keyboard;
+};
+
+struct tui_widget_ops {
+	TEE_Result (*paint)(struct tui_widget *tw, struct tui_state *ts);
+	void (*focus)(struct tui_widget *tw, struct tui_state *ts);
+	void (*unfocus)(struct tui_widget *tw, struct tui_state *ts);
+	void (*click)(struct tui_widget *tw, struct tui_state *ts,
+		      ssize_t x, ssize_t y);
+	bool (*input)(struct tui_widget *tw, struct tui_state *ts,
+		      uint32_t key);
+	void (*destroy)(struct tui_widget *tw);
+};
+
+struct tui_widget {
+	size_t x;
+	size_t y;
+	size_t width;
+	size_t height;
+	bool updated;
+	const struct tui_widget_ops *ops;
+	TAILQ_ENTRY(tui_widget) link;
+};
+
+TAILQ_HEAD(tui_widget_head, tui_widget);
+
+/*
+ * All functions here are prefixed with __tui since they are global symbols
+ * even if not directly exported to the TA.
+ */
+
+TEE_Result __tui_pta_get_screen_info(size_t *width, size_t *height,
+				     size_t *width_dpi, size_t *height_dpi,
+				     uint32_t *bpp);
+TEE_Result __tui_pta_init_session(void);
+TEE_Result __tui_pta_close_session(void);
+TEE_Result __tui_pta_init_screen(uint32_t color);
+TEE_Result __tui_pta_set_screen_image(size_t xpos, size_t ypos,
+				      struct image *image);
+TEE_Result __tui_pta_display_screen(bool clear_input, size_t timeout,
+				    bool *is_timedout, bool *is_key,
+				    uint32_t *key, size_t *xpos, size_t *ypos);
+
+void __tui_widget_add(struct tui_widget_head *twh, struct tui_widget *tw);
+struct image *__tui_widget_image_alloc(struct tui_widget *tw, uint32_t color);
+TEE_Result __tui_widget_display(struct tui_widget *tw, struct tui_state *ts,
+				struct image *image);
+struct tui_widget *__tui_widget_find(struct tui_widget_head *twh, size_t xpos,
+				     size_t ypos);
+bool __tui_widget_op_dummy_input(struct tui_widget *tw, struct tui_state *ts,
+				 uint32_t key);
+void __tui_widget_op_unfocus(struct tui_widget *tw, struct tui_state *ts);
+
+TEE_Result __tui_button_create(struct tui_widget_head *twh,
+			       TEE_TUIButtonType btype,
+			       TEE_TUIScreenButtonInfo *binfo,
+			       const TEE_TUIButton *custom_button,
+			       enum button_pos bpos, size_t scr_width,
+			       size_t scr_height);
+
+TEE_Result __tui_keyboard_create(struct tui_widget_head *twh, size_t *ypos,
+				 size_t scr_width);
+void __tui_keyboard_select_layout(struct tui_state *ts);
+struct tui_widget *__tui_keyboard_find(struct tui_widget_head *twh);
+
+TEE_Result __tui_entry_fields_create(struct tui_widget_head *twh,
+				     size_t *ypos, size_t scr_width,
+				     TEE_TUIScreenButtonInfo *cbi,
+				     TEE_TUIEntryField *efs, size_t num_efs);
+TEE_TUIEntryFieldType __tui_entry_field_get_input_type(struct tui_widget *tw);
+struct tui_widget *__tui_entry_field_find_first(struct tui_widget_head *twh);
+bool __tui_entry_field_has_temp_clear(struct tui_widget *tw);
+void __tui_entry_field_reset_temp_clear(struct tui_widget *tw);
+bool __tui_entry_fields_are_ok(struct tui_widget_head *twh);
+
+#endif /*__TUI_PRIVATE_H*/
+
diff --git a/lib/libutee/tui/tui_pta.c b/lib/libutee/tui/tui_pta.c
new file mode 100644
index 000000000..aeeb3f953
--- /dev/null
+++ b/lib/libutee/tui/tui_pta.c
@@ -0,0 +1,137 @@
+// SPDX-License-Identifier: BSD-2-Clause
+/*
+ * Copyright (c) 2016-2021, Linaro Limited
+ */
+
+#include <tee_api.h>
+#include <tee_tui_api.h>
+#include <pta_tui.h>
+#include <string.h>
+#include "image.h"
+#include "tui_private.h"
+
+static TEE_Result invoke_tui_pta(uint32_t cmd_id, uint32_t param_types,
+				 TEE_Param params[TEE_NUM_PARAMS])
+{
+	static TEE_TASessionHandle sess = TEE_HANDLE_NULL;
+	static const TEE_UUID core_uuid = PTA_TUI_UUID;
+
+	if (sess == TEE_HANDLE_NULL) {
+		TEE_Result res = TEE_OpenTASession(&core_uuid, 0, 0, NULL,
+						   &sess, NULL);
+
+		if (res != TEE_SUCCESS)
+			return res;
+	}
+
+	return TEE_InvokeTACommand(sess, 0, cmd_id, param_types, params, NULL);
+}
+
+TEE_Result __tui_pta_get_screen_info(size_t *width, size_t *height,
+				     size_t *width_dpi, size_t *height_dpi,
+				     uint32_t *bpp)
+{
+	TEE_Result res;
+	uint32_t param_types;
+	TEE_Param params[TEE_NUM_PARAMS];
+
+	param_types = TEE_PARAM_TYPES(TEE_PARAM_TYPE_VALUE_OUTPUT,
+				      TEE_PARAM_TYPE_VALUE_OUTPUT,
+				      TEE_PARAM_TYPE_VALUE_OUTPUT,
+				      TEE_PARAM_TYPE_NONE);
+	memset(params, 0, sizeof(params));
+	res = invoke_tui_pta(PTA_TUI_GET_SCREEN_INFO, param_types, params);
+	*width = params[0].value.a;
+	*height = params[0].value.b;
+	if (width_dpi)
+		*width_dpi = params[1].value.a;
+	if (height_dpi)
+		*height_dpi = params[1].value.b;
+	if (bpp)
+		*bpp = params[2].value.a;
+	return res;
+}
+
+TEE_Result __tui_pta_init_session(void)
+{
+	uint32_t param_types;
+	TEE_Param params[TEE_NUM_PARAMS];
+
+	param_types = TEE_PARAM_TYPES(TEE_PARAM_TYPE_NONE,
+				      TEE_PARAM_TYPE_NONE,
+				      TEE_PARAM_TYPE_NONE,
+				      TEE_PARAM_TYPE_NONE);
+	memset(params, 0, sizeof(params));
+	return invoke_tui_pta(PTA_TUI_INIT_SESSION, param_types, params);
+}
+
+TEE_Result __tui_pta_close_session(void)
+{
+	uint32_t param_types;
+	TEE_Param params[TEE_NUM_PARAMS];
+
+	param_types = TEE_PARAM_TYPES(TEE_PARAM_TYPE_NONE,
+				      TEE_PARAM_TYPE_NONE,
+				      TEE_PARAM_TYPE_NONE,
+				      TEE_PARAM_TYPE_NONE);
+	memset(params, 0, sizeof(params));
+	return invoke_tui_pta(PTA_TUI_CLOSE_SESSION, param_types, params);
+}
+
+TEE_Result __tui_pta_init_screen(uint32_t color)
+{
+	uint32_t param_types;
+	TEE_Param params[TEE_NUM_PARAMS];
+
+	param_types = TEE_PARAM_TYPES(TEE_PARAM_TYPE_VALUE_INPUT,
+				      TEE_PARAM_TYPE_NONE,
+				      TEE_PARAM_TYPE_NONE,
+				      TEE_PARAM_TYPE_NONE);
+	memset(params, 0, sizeof(params));
+	params[0].value.a = color;
+	return invoke_tui_pta(PTA_TUI_INIT_SCREEN, param_types, params);
+}
+
+TEE_Result __tui_pta_set_screen_image(size_t xpos, size_t ypos,
+				      struct image *image)
+{
+	uint32_t param_types;
+	TEE_Param params[TEE_NUM_PARAMS];
+
+	param_types = TEE_PARAM_TYPES(TEE_PARAM_TYPE_VALUE_INPUT,
+				      TEE_PARAM_TYPE_VALUE_INPUT,
+				      TEE_PARAM_TYPE_MEMREF_INPUT,
+				      TEE_PARAM_TYPE_NONE);
+	memset(params, 0, sizeof(params));
+	params[0].value.a = xpos;
+	params[0].value.b = ypos;
+	params[1].value.a = image->width;
+	params[1].value.b = image->height;
+	params[2].memref.buffer = image->buf;
+	params[2].memref.size = image->blen;
+	return invoke_tui_pta(PTA_TUI_SET_SCREEN_IMAGE, param_types, params);
+}
+
+TEE_Result __tui_pta_display_screen(bool clear_input, size_t timeout,
+				    bool *is_timedout, bool *is_key,
+				    uint32_t *key, size_t *xpos, size_t *ypos)
+{
+	TEE_Result res;
+	uint32_t param_types;
+	TEE_Param params[TEE_NUM_PARAMS];
+
+	param_types = TEE_PARAM_TYPES(TEE_PARAM_TYPE_VALUE_INPUT,
+				      TEE_PARAM_TYPE_VALUE_OUTPUT,
+				      TEE_PARAM_TYPE_VALUE_OUTPUT,
+				      TEE_PARAM_TYPE_VALUE_OUTPUT);
+	memset(params, 0, sizeof(params));
+	params[0].value.a = clear_input;
+	params[0].value.b = timeout;
+	res = invoke_tui_pta(PTA_TUI_DISPLAY_SCREEN, param_types, params);
+	*is_timedout = params[1].value.a;
+	*is_key = params[2].value.a;
+	*key = params[2].value.b;
+	*xpos = params[3].value.a;
+	*ypos = params[3].value.b;
+	return res;
+}
diff --git a/lib/libutee/tui/tui_widget.c b/lib/libutee/tui/tui_widget.c
new file mode 100644
index 000000000..12d0a4f6c
--- /dev/null
+++ b/lib/libutee/tui/tui_widget.c
@@ -0,0 +1,199 @@
+// SPDX-License-Identifier: BSD-2-Clause
+/*
+ * Copyright (c) 2016-2021, Linaro Limited
+ */
+
+#include <tee_api.h>
+#include <tee_tui_api.h>
+#include <stdlib.h>
+#include "font.h"
+#include "image.h"
+#include "tui_private.h"
+
+struct image *__tui_widget_image_alloc(struct tui_widget *tw, uint32_t color)
+{
+	return image_alloc(tw->width, tw->height, color);
+}
+
+TEE_Result __tui_widget_display(struct tui_widget *tw, struct tui_state *ts,
+				struct image *image)
+{
+	if (tw == ts->focus)
+		image_set_border(image, BORDER_SIZE, COLOR_RGB_RED);
+
+	return __tui_pta_set_screen_image(tw->x, tw->y, image);
+}
+
+bool __tui_widget_op_dummy_input(struct tui_widget *tw __unused,
+				 struct tui_state *ts __unused,
+				 uint32_t key __unused)
+{
+	EMSG("Ignoring key %#" PRIX32, key);
+	return false;
+}
+
+void __tui_widget_op_unfocus(struct tui_widget *tw, struct tui_state *ts)
+{
+	if (tw == ts->focus) {
+		tw->updated = true;
+		ts->focus = NULL;
+	}
+}
+
+void __tui_widget_add(struct tui_widget_head *twh, struct tui_widget *tw)
+{
+	struct tui_widget *t;
+
+	TAILQ_FOREACH(t, twh, link) {
+		if (tw->y < t->y || (tw->y == t->y && tw->x < t->y)) {
+			TAILQ_INSERT_BEFORE(t, tw, link);
+			return;
+		}
+	}
+	TAILQ_INSERT_TAIL(twh, tw, link);
+}
+
+struct tui_widget *__tui_widget_find(struct tui_widget_head *twh, size_t xpos,
+				     size_t ypos)
+{
+	struct tui_widget *tw;
+
+	TAILQ_FOREACH(tw, twh, link)
+		if (xpos >= tw->x && xpos < (tw->x + tw->width) &&
+		    ypos >= tw->y && ypos < (tw->y + tw->height))
+			return tw;
+	return NULL;
+}
+
+/*
+ * Button implementation.
+ */
+
+struct tui_button {
+	TEE_TUIButtonType type;
+	char *text;
+	struct tui_widget tw;
+};
+
+static bool __maybe_unused is_tui_button(struct tui_widget *tw);
+
+static struct tui_button *to_tui_button(struct tui_widget *tw)
+{
+	assert(is_tui_button(tw));
+	return container_of(tw, struct tui_button, tw);
+}
+
+static void op_button_focus(struct tui_widget *tw, struct tui_state *ts)
+{
+	if (tw != ts->focus) {
+		tw->updated = true;
+		ts->focus = tw;
+	}
+}
+
+static TEE_Result op_button_paint(struct tui_widget *tw, struct tui_state *ts)
+{
+	TEE_Result res;
+	struct tui_button *b = to_tui_button(tw);
+	struct image *image;
+
+	if (!tw->updated)
+		return TEE_SUCCESS;
+
+	image = __tui_widget_image_alloc(tw, COLOR_RGB_LIGHT_GOLDENROD);
+	if (!image)
+		return TEE_ERROR_OUT_OF_MEMORY;
+
+	if (!font_render_text(image, 0, 0, b->text, COLOR_RGB_BLACK)) {
+		res = TEE_ERROR_BAD_FORMAT;
+		goto out;
+	}
+	res = __tui_widget_display(tw, ts, image);
+	if (res == TEE_SUCCESS)
+		tw->updated = false;
+out:
+	image_free(image);
+	return res;
+}
+
+static void op_button_click(struct tui_widget *tw, struct tui_state *ts,
+			    ssize_t x __unused, ssize_t y __unused)
+{
+	struct tui_button *b = to_tui_button(tw);
+
+	ts->clicked_button = &b->type;
+	if (tw != ts->focus) {
+		ts->focus->ops->unfocus(ts->focus, ts);
+		tw->ops->focus(tw, ts);
+	}
+}
+
+static void op_button_destroy(struct tui_widget *tw)
+{
+	free(to_tui_button(tw));
+}
+
+static const struct tui_widget_ops button_ops = {
+	.paint = op_button_paint,
+	.focus = op_button_focus,
+	.unfocus = __tui_widget_op_unfocus,
+	.click = op_button_click,
+	.input = __tui_widget_op_dummy_input,
+	.destroy = op_button_destroy,
+};
+
+static bool is_tui_button(struct tui_widget *tw)
+{
+	return tw->ops == &button_ops;
+}
+
+TEE_Result __tui_button_create(struct tui_widget_head *twh,
+			       TEE_TUIButtonType btype,
+			       TEE_TUIScreenButtonInfo *binfo,
+			       const TEE_TUIButton *custom_button __unused,
+			       enum button_pos bpos,
+			       size_t scr_width, size_t scr_height)
+{
+	TEE_Result res;
+	struct tui_button *b = calloc(1, sizeof(*b));
+	size_t width;
+	size_t height;
+
+	if (!b)
+		return TEE_ERROR_OUT_OF_MEMORY;
+
+	b->tw.ops = &button_ops;
+	b->tw.updated = true;
+	b->type = btype;
+
+	b->text = binfo->buttonText;
+	if (!font_check_text_format(b->text, &width, &height, NULL)) {
+		res = TEE_ERROR_GENERIC;
+		goto out;
+	}
+
+	b->tw.width = width + BORDER_SIZE * 2;
+	b->tw.height = height + BORDER_SIZE * 2;
+	b->tw.y = scr_height - b->tw.height;
+	switch (bpos) {
+	case POS_LEFT:
+		b->tw.x = 0;
+		break;
+	case POS_MIDDLE:
+		b->tw.x = (scr_width / 2) - (b->tw.width / 2);
+		break;
+	case POS_RIGHT:
+		b->tw.x = scr_width - b->tw.width;
+		break;
+	default:
+		res = TEE_ERROR_GENERIC;
+		goto out;
+	}
+
+	res = TEE_SUCCESS;
+	__tui_widget_add(twh, &b->tw);
+out:
+	if (res != TEE_SUCCESS)
+		free(b);
+	return res;
+}
-- 
2.47.2

